File: 725. Split Linked List in Parts.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func splitListToParts(head *ListNode, k int) []*ListNode {
    count := getCount(head)
    n, remain, index := count/k, count%k, 0
    result := make([]*ListNode, k, k)
    for head != nil {
        result[index] = head
        index++
        group := 0
        if remain > 0 {
            group = n+1
            remain--
        } else {
            group = n
        }
        for group > 1 {
            head = head.Next
            group--
        }
        head, head.Next = head.Next, nil
    }
    return result
}

func getCount(head *ListNode) int {
    count := 0
    for head != nil {
        count++
        head = head.Next
    }
    return count
}

File: java/725. Split Linked List in Parts.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        int count = getCount(head);
        int n = count / k;
        int remain = count % k;
        ListNode[] answer = new ListNode[k];
        ListNode[] pointers = new ListNode[k];
        int index = 0;
        while (head != null) {
            int group = remain-- > 0 ? n+1 : n;
            answer[index] = head;
            while (group-- > 1) {
                head = head.next;
            }
            ListNode next = head.next;
            head.next = null;
            head = next;
            index++;
        }
        return answer;
    }
    private int getCount(ListNode head) {
        int count = 0;
        while (head != null) {
            head = head.next;
            count++;
        }
        return count;
    }
}

